### PR TITLE
adding fuzzers for libra

### PR DESCRIPTION
**Pull request by mimoo**

### CODE REVIEW

File: projects/libra/build.sh [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    zip -r $OUT/"$line"_seed_corpus.zip fuzz/corpus/$line
+done
+
+# build fuzzers
+cat fuzzer_list | while read -r line
+do
+    # --cfg fuzzing      -> used to change code logic
+    # -Cdebug-assertions -> to get debug_assert in rust
+    # other flags        -> taken from cargo fuzz
+    export RUSTFLAGS="--cfg fuzzing -Cpasses=sancov -Cllvm-args=-sanitizer-coverage-level=4 -Cllvm-args=-sanitizer-coverage-trace-compares -Cllvm-args=-sanitizer-coverage-inline-8bit-counters -Cllvm-args=-sanitizer-coverage-trace-geps -Cllvm-args=-sanitizer-coverage-prune-blocks=0 -Cllvm-args=-sanitizer-coverage-pc-table -Clink-dead-code -Cllvm-args=-sanitizer-coverage-stack-depth -Zsanitizer=address -Cdebug-assertions -Ccodegen-units=1"
```
Comment (jonathanmetzman) [inline projects/libra/build.sh]: I think stack depth flag is only useful if you *want* to find stack overflows (not stack buffer overflows).
@kcc?

File: projects/libra/build.sh [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+export CXXFLAGS_EXTRA="-stdlib=libc++"
+export CXXFLAGS="$CFLAGS $CXXFLAGS_EXTRA"
+
+# RUSTC_BOOTSTRAP: to get some nightly features like ASAN
+export RUSTC_BOOTSTRAP=1 
+# ASAN options:
+# - detect_odr_violation=0 - for address sanitizer
+# - detect_leaks=1 - will detect leaks as well
+# - symbolize=1 - better stack trace in case of bugs
+export ASAN_OPTIONS=detect_odr_violation=0:detect_leaks=1:symbolize=1
```
Comment (jonathanmetzman) [inline projects/libra/build.sh]: This is supposed to be used during compilation not while running the fuzzers right?

**Comment by jonathanmetzman** [comment]
I don't think AFL is supported by rust projects and I'm not sure UBSAN is useful or even works. Probably best to disable both. Will merge after that

**Comment by jonathanmetzman** [comment]
I don't think AFL is supported by rust projects and I'm not sure UBSAN is useful or even works. 
These failed in CI.
Probably best to disable both. Will merge after that

File: projects/libra/build.sh (hunk 2 above)
Comment (mimoo) [reply projects/libra/build.sh]: indeed!

File: projects/libra/build.sh (hunk 1 above)
Comment (mimoo) [reply projects/libra/build.sh]: I can't find any documentation on `sanitizer-coverage-stack-depth`, I copied this argument and others from [cargo fuzz](https://github.com/rust-fuzz/cargo-fuzz/).

**Comment by mimoo** [comment]
> Probably best to disable both.

How do I do that : o ?

**Comment by jonathanmetzman** [comment]
Add this to your project.yaml:
```
sanitizers:
  - address
fuzzing_engines:
  - libfuzzer
language: rust
```

File: projects/libra/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+# limitations under the License.
+#
+################################################################################
+
+# google oss-fuzz stuff
+FROM gcr.io/oss-fuzz-base/base-builder
+MAINTAINER davidwg@fb.com
+
+# install other tools we might need
+# RUN apt-get update && apt-get install -y make autoconf automake libtool curl cmake python llvm-dev
```
Comment (Dor1s) [inline projects/libra/Dockerfile]: nit: can we delete this and lines 27, 29?

File: projects/libra/project.yaml [hunk 4]
```diff
@@ -2,3 +2,7 @@ homepage: "https://github.com/libra/libra"
 primary_contact: "davidwong.crypto@gmail.com"
 auto_ccs:
   - "davidwg@fb.com"
+sanitizers:
+  - address
+fuzzing_engines:
+  - libfuzzer
```
Comment (Dor1s) [inline projects/libra/project.yaml]: +1 please add `language: rust`, and we should think whether we can automatically catch non C/C++ projects without `language` attribute

**Comment by mimoo** [comment]
done!

**Comment by mimoo** [comment]
> The job exceeded the maximum time limit for jobs, and has been terminated.

Is it normal that this is taking so much time? What will happen when we add more fuzzers?

**Comment by Dor1s** [comment]
> Is it normal that this is taking so much time? What will happen when we add more fuzzers?

Yes, we have a handful of large projects that Travis fails to test in under 3 hours (a limit that we can't increase any more, I believe).

So far we used to just go ahead for such failures. Our real build infra is more powerful and can run longer. If any actual build or test failure happens there, it will automatically report a bug to you.

What's really impressive is that the latest Travis attempt took ` Ran for 2 hrs 59 min 55 sec`, which is very slightly under 3 hours :D

**Comment by mimoo** [comment]
got lucky :D

**Pull request merged**